src/Microsoft.AspNetCore.Routing/Matching/DfaMatcherBuilder.cs

```diff
@@ -145,72 +147,80 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
                         var part = segment.Parts[0];
                         if (segment.IsSimple && part is RoutePatternLiteralPart literalPart)
                         {
-                            DfaNode next = null;
-                            var literal = literalPart.Content;
-                            if (parent.Literals == null ||
-                                !parent.Literals.TryGetValue(literal, out next))
-                            {
-                                next = new DfaNode()
-                                {
-                                    PathDepth = parent.PathDepth + 1,
-                                    Label = includeLabel ? parent.Label + literal + "/" : null,
-                                };
-                                parent.AddLiteral(literal, next);
-                            }
-
-                            nextParents.Add(next);
+                            AddLiteralNode(includeLabel, nextParents, parent, literalPart.Content);
                         }
-                        else if (segment.IsSimple && part is RoutePatternParameterPart parameterPart && parameterPart.IsCatchAll)
+                        else if (segment.IsSimple && part is RoutePatternParameterPart parameterPart)
                         {
-                            // A catch all should traverse all literal nodes as well as parameter nodes
-                            // we don't need to create the parameter node here because of ordering
-                            // all catchalls will be processed after all parameters.
-                            if (parent.Literals != null)
-                            {
-                                nextParents.AddRange(parent.Literals.Values);
-                            }
-                            if (parent.Parameters != null)
-                            {
-                                nextParents.Add(parent.Parameters);
-                            }
-
-                            // We also create a 'catchall' here. We don't do further traversals
-                            // on the catchall node because only catchalls can end up here. The
-                            // catchall node allows us to capture an unlimited amount of segments
-                            // and also to match a zero-length segment, which a parameter node
-                            // doesn't allow.
-                            if (parent.CatchAll == null)
+                            if (parameterPart.IsCatchAll)
```

Comment:
Why move just this part into a separate if? this seems really non-obvious

Reply:
Because there are now multiple actions to take for parameter parts:
* Is it a catch all
* Does it have a required value
* Everything else

The alternative to nesting is replicating the check that the part is a parameter in the parent if else and that seemed less intuitive to me.

Reply:
That seems great to me.